File: src/main/java/com/fish/extendedae_plus_client/widgets/button/EAEPCycleButton.java
package com.fish.extendedae_plus_client.widgets.button;

import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.neoforged.neoforge.network.PacketDistributor;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.IntUnaryOperator;

public class EAEPCycleButton extends EAEPButton {
    protected final List<EAEPActionItems> states;
    protected final IntUnaryOperator stateIterator;
    protected int stateIndex = 0;

    public EAEPCycleButton(List<EAEPActionItems> states,
                           BiConsumer<Integer, EAEPActionItems> statedOnPress,
                           @Nullable IntUnaryOperator stateIterator) {
        super(button -> {
            if (!(button instanceof EAEPCycleButton cycleButton)) return;
            statedOnPress.accept(cycleButton.iterateState(), cycleButton.getAction());
        });

        this.states = states;
        this.stateIterator = stateIterator;
    }

    @Override
    public EAEPActionItems getAction() {
        if (this.states == null) return EAEPActionItems.BACKING_OUT;
        return this.states.get(this.stateIndex);
    }

    public void setStateIndex(int stateIndex) {
        this.setStateIndex(stateIndex, false);
    }

    public void setStateIndex(int stateIndex, boolean triggerEvent) {
        this.stateIndex = stateIndex;
        if (triggerEvent) this.onPress();
        else this.updateTooltip();
    }

    public int getStateIndex() {
        return this.stateIndex;
    }

    /// @return 被迭代过的 stateIndex
    public int iterateState() {
        if (this.stateIterator != null)
            this.stateIndex = this.stateIterator.applyAsInt(this.stateIndex);
        else this.stateIndex = (this.stateIndex + 1) % this.states.size();
        return this.stateIndex;
    }

    public static final class Builder {
        private final List<EAEPActionItems> states = new ArrayList<>();
        private final List<Consumer<EAEPActionItems>> tasks = new ArrayList<>();
        private IntUnaryOperator stateIterator = null;

        public Builder addPart(EAEPActionItems action) {
            return this.addPart(action, ignored -> {});
        }

        public Builder addPart(EAEPActionItems action, CustomPacketPayload packet) {
            return this.addPart(action, () -> PacketDistributor.sendToServer(packet));
        }

        public Builder addPart(EAEPActionItems action, Runnable onPress) {
            return this.addPart(action, ignored -> onPress.run());
        }

        public Builder addPart(EAEPActionItems action, @Nullable Consumer<EAEPActionItems> onPress) {
            if (onPress == null) onPress = ignored -> {};

            this.states.add(action);
            this.tasks.add(onPress);
            return this;
        }

        public Builder setIterator(IntUnaryOperator stateIterator) {
            this.stateIterator = stateIterator;
            return this;
        }

        public EAEPCycleButton build() {
            return new EAEPCycleButton(this.states,
                    (index, action) -> this.tasks.get(index).accept(action),
                    this.stateIterator);
        }
    }
}

File: src/main/kotlin/com/fish/extendedae_plus_client/render/widgets/button/EAEPCycleButton.kt
package com.fish.extendedae_plus_client.render.widgets.button

import appeng.client.gui.AEBaseScreen
import net.minecraft.client.Minecraft
import java.util.function.BiConsumer
import java.util.function.Consumer

open class EAEPCycleButton(
    protected val states: List<EAEPActionItems>,
    statedTask: BiConsumer<Int, EAEPActionItems>,
    protected val iteratorState: IteratorState
) : EAEPButton({ button ->
    val screen = Minecraft.getInstance().screen
    if (button is EAEPCycleButton && screen is AEBaseScreen<*>)
        statedTask.accept(button.iterateState(screen.isHandlingRightClick), button.action)
}) {
    protected var stateIndex: Int = 0

    override val action: EAEPActionItems
        get() = this.states[this.stateIndex]

    fun setStateIndex(stateIndex: Int, triggerEvent: Boolean) {
        this.stateIndex = stateIndex
        if (triggerEvent) this.onPress()
        else this.updateTooltip()
    }

    /**
     * @return 被迭代过的 stateIndex
     */
    fun iterateState(reversed: Boolean): Int {
        this.iteratorState.let { this.stateIndex = it.iterate(this.stateIndex, reversed) }
        return this.stateIndex
    }

    class Builder {
        private val states: MutableList<EAEPActionItems> = ArrayList()
        private val tasks: MutableList<Consumer<EAEPActionItems>> = ArrayList()
        private var task: Consumer<EAEPActionItems>? = null
        private var iteratorState: IteratorState? = null

        fun addPart(action: EAEPActionItems, onPress: Runnable): Builder {
            return this.addPart(action) { _ -> onPress.run() }
        }

        fun addPart(action: EAEPActionItems, onPress: Consumer<EAEPActionItems>? = Consumer {}): Builder {
            this.states.add(action)
            this.tasks.add(onPress?: Consumer {})
            return this
        }

        fun addGlobalTask(task: Consumer<EAEPActionItems>?): Builder {
            this.task = task
            return this
        }

        fun setIterator(iteratorState: IteratorState): Builder {
            this.iteratorState = iteratorState
            return this
        }

        fun build(): EAEPCycleButton {
            return EAEPCycleButton(
                this.states,
                { index: Int, action: EAEPActionItems ->
                    this.task?.accept(action)
                    this.tasks[index].accept(action)
                },
                this.iteratorState ?: IteratorState { prev, reversed ->
                    (prev + if (reversed) -1 else 1 + this.states.size) % this.states.size
                }
            )
        }
    }

    fun interface IteratorState {
        fun iterate(prev: Int, reversed: Boolean): Int
    }
}
